File: statCards.py
"""Resolution-independent stat symbols and collectible mini upgrade cards."""

import pygame as pg

import uiTheme as ui


def _line(surface, color, start, end, width):
    pg.draw.line(surface, color, start, end, max(1, int(width)))


def draw_stat_symbol(surface, stat_name, rect, color=ui.TEXT):
    """Draw a compact silhouette for every upgrade stat using Pygame primitives."""
    rect = pg.Rect(rect)
    cx, cy = rect.center
    unit = max(1, min(rect.width, rect.height) / 20)
    stroke = max(1, round(unit * 1.7))
    name = stat_name.lower()

    def bullet(x=cx, y=cy, scale=1.0):
        length, radius = 10 * unit * scale, 3.5 * unit * scale
        body = pg.Rect(x - length * .45, y - radius, length * .65, radius * 2)
        pg.draw.rect(surface, color, body, border_radius=max(1, round(radius)))
        pg.draw.polygon(surface, color, ((body.right - unit, body.top),
                                         (x + length * .55, y),
                                         (body.right - unit, body.bottom)))

    if name == "defense":
        points = ((cx, cy - 8 * unit), (cx + 7 * unit, cy - 5 * unit),
                  (cx + 6 * unit, cy + 3 * unit), (cx, cy + 9 * unit),
                  (cx - 6 * unit, cy + 3 * unit), (cx - 7 * unit, cy - 5 * unit))
        pg.draw.polygon(surface, color, points, stroke)
        _line(surface, color, (cx, cy - 6 * unit), (cx, cy + 6 * unit), stroke)
    elif name in ("health", "vitality"):
        _line(surface, color, (cx - 7 * unit, cy), (cx + 7 * unit, cy), stroke)
        _line(surface, color, (cx, cy - 7 * unit), (cx, cy + 7 * unit), stroke)
        if name == "vitality":
            pg.draw.circle(surface, color, (cx, cy), round(9 * unit), stroke)
    elif name == "bullet pierce":
        bullet(cx - 2 * unit, cy)
        for x in (cx + 5 * unit, cx + 8 * unit):
            _line(surface, color, (x, cy - 7 * unit), (x, cy + 7 * unit), stroke)
    elif name == "bullet count":
        for offset in (-5, 0, 5):
            bullet(cx - 1 * unit, cy + offset * unit, .72)
    elif name == "spread angle":
        origin = (cx - 7 * unit, cy)
        for offset in (-7, 0, 7):
            _line(surface, color, origin, (cx + 8 * unit, cy + offset * unit), stroke)
    elif name == "attack speed":
        pg.draw.circle(surface, color, (cx, cy), round(8 * unit), stroke)
        _line(surface, color, (cx, cy), (cx + 5 * unit, cy - 4 * unit), stroke)
        for offset in (-4, 0, 4):
            _line(surface, color, (cx - 11 * unit, cy + offset * unit),
                  (cx - 7 * unit, cy + offset * unit), stroke)
    elif name == "bullet speed":
        bullet(cx + 2 * unit, cy)
        for offset, length in ((-5, 5), (0, 8), (5, 5)):
            _line(surface, color, (cx - (length + 3) * unit, cy + offset * unit),
                  (cx - 3 * unit, cy + offset * unit), stroke)
    elif name == "bullet range":
        bullet(cx - 4 * unit, cy)
        _line(surface, color, (cx + 2 * unit, cy), (cx + 9 * unit, cy), stroke)
        _line(surface, color, (cx + 7 * unit, cy - 3 * unit), (cx + 10 * unit, cy), stroke)
        _line(surface, color, (cx + 7 * unit, cy + 3 * unit), (cx + 10 * unit, cy), stroke)
    elif name == "bullet damage":
        bullet(cx - 3 * unit, cy)
        for angle in range(0, 360, 45):
            direction = pg.Vector2(1, 0).rotate(angle)
            _line(surface, color, (cx + direction.x * 4 * unit, cy + direction.y * 4 * unit),
                  (cx + direction.x * 8 * unit, cy + direction.y * 8 * unit), stroke)
    elif name == "bullet size":
        bullet(cx - 2 * unit, cy + 2 * unit, 1.35)
        _line(surface, color, (cx + 7 * unit, cy + 6 * unit), (cx + 7 * unit, cy - 7 * unit), stroke)
        pg.draw.polygon(surface, color, ((cx + 7 * unit, cy - 9 * unit),
                                         (cx + 3 * unit, cy - 4 * unit),
                                         (cx + 11 * unit, cy - 4 * unit)))
    elif name == "player speed":
        pg.draw.circle(surface, color, (cx + 2 * unit, cy - 6 * unit), round(2.5 * unit))
        _line(surface, color, (cx + 1 * unit, cy - 3 * unit), (cx - 2 * unit, cy + 2 * unit), stroke)
        _line(surface, color, (cx - 2 * unit, cy + 2 * unit), (cx + 4 * unit, cy + 7 * unit), stroke)
        _line(surface, color, (cx - 2 * unit, cy + 2 * unit), (cx - 7 * unit, cy + 8 * unit), stroke)
        for offset in (-5, 0, 5):
            _line(surface, color, (cx - 10 * unit, cy + offset * unit),
                  (cx - 6 * unit, cy + offset * unit), stroke)
    elif name == "crit chance":
        pg.draw.circle(surface, color, (cx, cy), round(8 * unit), stroke)
        pg.draw.circle(surface, color, (cx, cy), round(3 * unit), stroke)
        _line(surface, color, (cx + 2 * unit, cy - 2 * unit),
              (cx + 9 * unit, cy - 9 * unit), stroke)
    elif name == "crit damage":
        points = []
        for index in range(16):
            radius = (9 if index % 2 == 0 else 4) * unit
            point = pg.Vector2(0, -radius).rotate(index * 22.5)
            points.append((cx + point.x, cy + point.y))
        pg.draw.polygon(surface, color, points, stroke)
    elif name == "aura size":
        for radius in (3, 6, 9):
            pg.draw.circle(surface, color, (cx, cy), round(radius * unit), stroke)
    elif name == "aura strength":
        pg.draw.circle(surface, color, (cx, cy), round(3 * unit), stroke)
        for angle in (0, 90, 180, 270):
            direction = pg.Vector2(0, -1).rotate(angle)
            start = (cx + direction.x * 9 * unit, cy + direction.y * 9 * unit)
            end = (cx + direction.x * 4 * unit, cy + direction.y * 4 * unit)
            _line(surface, color, start, end, stroke)
            side = direction.rotate(28)
            pg.draw.polygon(surface, color, (end,
                (end[0] + side.x * 3 * unit, end[1] + side.y * 3 * unit),
                (end[0] + direction.rotate(-28).x * 3 * unit,
                 end[1] + direction.rotate(-28).y * 3 * unit)))
    elif name == "exp multiplier":
        gem = ((cx, cy - 9 * unit), (cx + 7 * unit, cy - 3 * unit),
               (cx + 5 * unit, cy + 7 * unit), (cx - 5 * unit, cy + 7 * unit),
               (cx - 7 * unit, cy - 3 * unit))
        pg.draw.polygon(surface, color, gem, stroke)
        _line(surface, color, (cx - 6 * unit, cy - 3 * unit),
              (cx + 6 * unit, cy - 3 * unit), stroke)
    else:
        pg.draw.circle(surface, color, (cx, cy), round(7 * unit), stroke)
        _line(surface, color, (cx, cy - 4 * unit), (cx, cy + 4 * unit), stroke)


def draw_upgrade_card(surface, rect, stat_name, rarity, math_type, hovered=False):
    """Draw a rarity-backed mini card with a stat icon and scaling corner mark."""
    rect = pg.Rect(rect)
    rarity_color = ui.RARITY_COLORS.get(rarity, ui.BORDER)
    shadow = rect.move(max(2, rect.width // 12), max(2, rect.width // 12))
    pg.draw.rect(surface, ui.SHADOW, shadow, border_radius=max(2, rect.width // 8))
    fill = ui.lighten(rarity_color, 24) if hovered else rarity_color
    pg.draw.rect(surface, fill, rect, border_radius=max(2, rect.width // 8))
    pg.draw.rect(surface, ui.INK, rect, max(2, rect.width // 14),
                 border_radius=max(2, rect.width // 8))
    inner = rect.inflate(-rect.width * .18, -rect.height * .22)
    draw_stat_symbol(surface, stat_name, inner, ui.INK)

    mark_center = (rect.right - rect.width * .18, rect.y + rect.height * .17)
    mark = rect.width * .10
    stroke = max(2, rect.width // 14)
    if math_type == "multiplicative":
        _line(surface, ui.TEXT, (mark_center[0] - mark, mark_center[1] - mark),
              (mark_center[0] + mark, mark_center[1] + mark), stroke)
        _line(surface, ui.TEXT, (mark_center[0] + mark, mark_center[1] - mark),
              (mark_center[0] - mark, mark_center[1] + mark), stroke)
    else:
        _line(surface, ui.TEXT, (mark_center[0] - mark, mark_center[1]),
              (mark_center[0] + mark, mark_center[1]), stroke)
        _line(surface, ui.TEXT, (mark_center[0], mark_center[1] - mark),
              (mark_center[0], mark_center[1] + mark), stroke)
    return rect

File: RotBoiRemastered/UI/StatCards.cs
using Microsoft.Xna.Framework;
using Microsoft.Xna.Framework.Graphics;
using RotBoiRemastered.Core;

namespace RotBoiRemastered.UI;

/// <summary>Resolution-independent stat symbols and collectible mini upgrade cards. Ported from statCards.py.</summary>
public static class StatCards
{
    private static readonly Dictionary<string, string> NormalizedStatNames = new(StringComparer.Ordinal);

    private static void DrawLine(SpriteBatch spriteBatch, Color color, Vector2 start, Vector2 end, float width)
        => Primitives2D.Line(spriteBatch, start, end, color, Math.Max(1, (int)width));

    private static string NormalizeStatName(string statName)
    {
        if (NormalizedStatNames.TryGetValue(statName, out string? normalized))
            return normalized;

        normalized = statName.ToLowerInvariant();
        NormalizedStatNames[statName] = normalized;
        return normalized;
    }

    private static Vector2 Rotate(Vector2 v, float degrees)
    {
        float radians = MathHelper.ToRadians(degrees);
        float cos = MathF.Cos(radians), sin = MathF.Sin(radians);
        return new Vector2(v.X * cos - v.Y * sin, v.X * sin + v.Y * cos);
    }

    /// <summary>Draw a compact silhouette for every upgrade stat using Primitives2D.</summary>
    public static void DrawStatSymbol(SpriteBatch spriteBatch, string statName, Rectangle rect, Color? color = null)
    {
        Color drawColor = color ?? UiTheme.Text;
        float cx = rect.Center.X, cy = rect.Center.Y;
        float unit = Math.Max(1f, Math.Min(rect.Width, rect.Height) / 20f);
        float stroke = Math.Max(1f, MathF.Round(unit * 1.7f));
        string name = NormalizeStatName(statName);

        void Bullet(float x, float y, float scale = 1.0f)
        {
            float length = 10 * unit * scale, radius = 3.5f * unit * scale;
            var body = new Rectangle((int)(x - length * .45f), (int)(y - radius), (int)(length * .65f), (int)(radius * 2));
            Primitives2D.FillRoundedRect(spriteBatch, body, drawColor, Math.Max(1, (int)MathF.Round(radius)));
            Span<Vector2> tip = stackalloc Vector2[]
            {
                new Vector2(body.Right - unit, body.Top),
                new Vector2(x + length * .55f, y),
                new Vector2(body.Right - unit, body.Bottom),
            };
            Primitives2D.FillPolygonSpan(spriteBatch, tip, drawColor);
        }

        switch (name)
        {
            case "defense":
            {
                Span<Vector2> points = stackalloc Vector2[]
                {
                    new Vector2(cx, cy - 8 * unit), new Vector2(cx + 7 * unit, cy - 5 * unit),
                    new Vector2(cx + 6 * unit, cy + 3 * unit), new Vector2(cx, cy + 9 * unit),
                    new Vector2(cx - 6 * unit, cy + 3 * unit), new Vector2(cx - 7 * unit, cy - 5 * unit),
                };
                Primitives2D.PolygonOutlineSpan(spriteBatch, points, drawColor, (int)stroke);
                DrawLine(spriteBatch, drawColor, new Vector2(cx, cy - 6 * unit), new Vector2(cx, cy + 6 * unit), stroke);
                break;
            }
            case "health" or "vitality":
                DrawLine(spriteBatch, drawColor, new Vector2(cx - 7 * unit, cy), new Vector2(cx + 7 * unit, cy), stroke);
                DrawLine(spriteBatch, drawColor, new Vector2(cx, cy - 7 * unit), new Vector2(cx, cy + 7 * unit), stroke);
                if (name == "vitality")
                    Primitives2D.CircleOutline(spriteBatch, new Vector2(cx, cy), MathF.Round(9 * unit), drawColor, (int)stroke);
                break;
            case "bullet pierce":
                Bullet(cx - 2 * unit, cy);
                DrawLine(spriteBatch, drawColor, new Vector2(cx + 5 * unit, cy - 7 * unit), new Vector2(cx + 5 * unit, cy + 7 * unit), stroke);
                DrawLine(spriteBatch, drawColor, new Vector2(cx + 8 * unit, cy - 7 * unit), new Vector2(cx + 8 * unit, cy + 7 * unit), stroke);
                break;
            case "bullet count":
                for (float offset = -5f; offset <= 5f; offset += 5f)
                    Bullet(cx - 1 * unit, cy + offset * unit, .72f);
                break;
            case "spread angle":
            {
                var origin = new Vector2(cx - 7 * unit, cy);
                for (float offset = -7f; offset <= 7f; offset += 7f)
                    DrawLine(spriteBatch, drawColor, origin, new Vector2(cx + 8 * unit, cy + offset * unit), stroke);
                break;
            }
            case "attack speed":
                Primitives2D.CircleOutline(spriteBatch, new Vector2(cx, cy), MathF.Round(8 * unit), drawColor, (int)stroke);
                DrawLine(spriteBatch, drawColor, new Vector2(cx, cy), new Vector2(cx + 5 * unit, cy - 4 * unit), stroke);
                for (float offset = -4f; offset <= 4f; offset += 4f)
                    DrawLine(spriteBatch, drawColor, new Vector2(cx - 11 * unit, cy + offset * unit), new Vector2(cx - 7 * unit, cy + offset * unit), stroke);
                break;
            case "bullet speed":
            {
                Bullet(cx + 2 * unit, cy);
                void SpeedLine(float offset, float length) =>
                    DrawLine(spriteBatch, drawColor, new Vector2(cx - (length + 3) * unit, cy + offset * unit), new Vector2(cx - 3 * unit, cy + offset * unit), stroke);
                SpeedLine(-5f, 5f);
                SpeedLine(0f, 8f);
                SpeedLine(5f, 5f);
                break;
            }
            case "bullet range":
                Bullet(cx - 4 * unit, cy);
                DrawLine(spriteBatch, drawColor, new Vector2(cx + 2 * unit, cy), new Vector2(cx + 9 * unit, cy), stroke);
                DrawLine(spriteBatch, drawColor, new Vector2(cx + 7 * unit, cy - 3 * unit), new Vector2(cx + 10 * unit, cy), stroke);
                DrawLine(spriteBatch, drawColor, new Vector2(cx + 7 * unit, cy + 3 * unit), new Vector2(cx + 10 * unit, cy), stroke);
                break;
            case "bullet damage":
                Bullet(cx - 3 * unit, cy);
                for (int angle = 0; angle < 360; angle += 45)
                {
                    var direction = Rotate(new Vector2(1, 0), angle);
                    DrawLine(spriteBatch, drawColor, new Vector2(cx + direction.X * 4 * unit, cy + direction.Y * 4 * unit),
                        new Vector2(cx + direction.X * 8 * unit, cy + direction.Y * 8 * unit), stroke);
                }
                break;
            case "bullet size":
                Bullet(cx - 2 * unit, cy + 2 * unit, 1.35f);
                DrawLine(spriteBatch, drawColor, new Vector2(cx + 7 * unit, cy + 6 * unit), new Vector2(cx + 7 * unit, cy - 7 * unit), stroke);
                Primitives2D.FillPolygonSpan(spriteBatch, stackalloc Vector2[]
                {
                    new Vector2(cx + 7 * unit, cy - 9 * unit),
                    new Vector2(cx + 3 * unit, cy - 4 * unit),
                    new Vector2(cx + 11 * unit, cy - 4 * unit),
                }, drawColor);
                break;
            case "player speed":
                Primitives2D.FillCircle(spriteBatch, new Vector2(cx + 2 * unit, cy - 6 * unit), MathF.Round(2.5f * unit), drawColor);
                DrawLine(spriteBatch, drawColor, new Vector2(cx + 1 * unit, cy - 3 * unit), new Vector2(cx - 2 * unit, cy + 2 * unit), stroke);
                DrawLine(spriteBatch, drawColor, new Vector2(cx - 2 * unit, cy + 2 * unit), new Vector2(cx + 4 * unit, cy + 7 * unit), stroke);
                DrawLine(spriteBatch, drawColor, new Vector2(cx - 2 * unit, cy + 2 * unit), new Vector2(cx - 7 * unit, cy + 8 * unit), stroke);
                for (float offset = -5f; offset <= 5f; offset += 5f)
                    DrawLine(spriteBatch, drawColor, new Vector2(cx - 10 * unit, cy + offset * unit), new Vector2(cx - 6 * unit, cy + offset * unit), stroke);
                break;
            case "crit chance":
                Primitives2D.CircleOutline(spriteBatch, new Vector2(cx, cy), MathF.Round(8 * unit), drawColor, (int)stroke);
                Primitives2D.CircleOutline(spriteBatch, new Vector2(cx, cy), MathF.Round(3 * unit), drawColor, (int)stroke);
                DrawLine(spriteBatch, drawColor, new Vector2(cx + 2 * unit, cy - 2 * unit), new Vector2(cx + 9 * unit, cy - 9 * unit), stroke);
                break;
            case "crit damage":
            {
                Span<Vector2> points = stackalloc Vector2[16];
                for (int index = 0; index < 16; index++)
                {
                    float radius = (index % 2 == 0 ? 9f : 4f) * unit;
                    var point = Rotate(new Vector2(0, -radius), index * 22.5f);
                    points[index] = new Vector2(cx + point.X, cy + point.Y);
                }
                Primitives2D.PolygonOutlineSpan(spriteBatch, points, drawColor, (int)stroke);
                break;
            }
            case "aura size":
                for (float radius = 3f; radius <= 9f; radius += 3f)
                    Primitives2D.CircleOutline(spriteBatch, new Vector2(cx, cy), MathF.Round(radius * unit), drawColor, (int)stroke);
                break;
            case "aura strength":
                Primitives2D.CircleOutline(spriteBatch, new Vector2(cx, cy), MathF.Round(3 * unit), drawColor, (int)stroke);
                Span<Vector2> arrow = stackalloc Vector2[3];
                for (float angle = 0f; angle < 360f; angle += 90f)
                {
                    var direction = Rotate(new Vector2(0, -1), angle);
                    var start = new Vector2(cx + direction.X * 9 * unit, cy + direction.Y * 9 * unit);
                    var end = new Vector2(cx + direction.X * 4 * unit, cy + direction.Y * 4 * unit);
                    DrawLine(spriteBatch, drawColor, start, end, stroke);
                    var sideA = Rotate(direction, 28);
                    var sideB = Rotate(direction, -28);
                    arrow[0] = end;
                    arrow[1] = new Vector2(end.X + sideA.X * 3 * unit, end.Y + sideA.Y * 3 * unit);
                    arrow[2] = new Vector2(end.X + sideB.X * 3 * unit, end.Y + sideB.Y * 3 * unit);
                    Primitives2D.FillPolygonSpan(spriteBatch, arrow, drawColor);
                }
                break;
            case "exp multiplier":
            {
                Span<Vector2> gem = stackalloc Vector2[]
                {
                    new Vector2(cx, cy - 9 * unit), new Vector2(cx + 7 * unit, cy - 3 * unit),
                    new Vector2(cx + 5 * unit, cy + 7 * unit), new Vector2(cx - 5 * unit, cy + 7 * unit),
                    new Vector2(cx - 7 * unit, cy - 3 * unit),
                };
                Primitives2D.PolygonOutlineSpan(spriteBatch, gem, drawColor, (int)stroke);
                DrawLine(spriteBatch, drawColor, new Vector2(cx - 6 * unit, cy - 3 * unit), new Vector2(cx + 6 * unit, cy - 3 * unit), stroke);
                break;
            }
            case "mind tokens":
                Primitives2D.CircleOutline(spriteBatch, new Vector2(cx, cy), MathF.Round(7 * unit), drawColor, (int)stroke);
                Primitives2D.CircleOutline(spriteBatch, new Vector2(cx, cy), MathF.Round(3.5f * unit), drawColor, Math.Max(1, (int)(stroke * .7f)));
                break;
            case "vault":
            {
                var body = new Rectangle((int)(cx - 6 * unit), (int)(cy - 2 * unit), (int)(12 * unit), (int)(9 * unit));
                Primitives2D.RectOutline(spriteBatch, body, drawColor, (int)stroke);
                Primitives2D.CircleOutline(spriteBatch, new Vector2(cx, cy - 2 * unit), MathF.Round(4 * unit), drawColor, (int)stroke);
                Primitives2D.FillRect(spriteBatch,
                    new Rectangle((int)(cx - unit), (int)(cy + .5f * unit), (int)(2 * unit), (int)(2 * unit)), drawColor);
                break;
            }
            case "quests":
            {
                DrawLine(spriteBatch, drawColor, new Vector2(cx - 6 * unit, cy + 8 * unit), new Vector2(cx - 6 * unit, cy - 8 * unit), stroke);
                Span<Vector2> flag = stackalloc Vector2[]
                {
                    new Vector2(cx - 6 * unit, cy - 8 * unit), new Vector2(cx + 6 * unit, cy - 5 * unit),
                    new Vector2(cx - 6 * unit, cy - 2 * unit),
                };
                Primitives2D.FillPolygonSpan(spriteBatch, flag, drawColor);
                break;
            }
            case "skills":
            {
                for (float angle = 0f; angle < 360f; angle += 72f)
                {
                    var direction = Rotate(new Vector2(0, -1), angle);
                    DrawLine(spriteBatch, drawColor, new Vector2(cx, cy),
                        new Vector2(cx + direction.X * 8 * unit, cy + direction.Y * 8 * unit), stroke);
                }
                Primitives2D.CircleOutline(spriteBatch, new Vector2(cx, cy), MathF.Round(2.5f * unit), drawColor, (int)stroke);
                break;
            }
            default:
                Primitives2D.CircleOutline(spriteBatch, new Vector2(cx, cy), MathF.Round(7 * unit), drawColor, (int)stroke);
                DrawLine(spriteBatch, drawColor, new Vector2(cx, cy - 4 * unit), new Vector2(cx, cy + 4 * unit), stroke);
                break;
        }
    }

    /// <summary>Draws a rarity-backed mini card with a stat icon and scaling corner mark.</summary>
    public static Rectangle DrawUpgradeCard(SpriteBatch spriteBatch, Rectangle rect, string statName, string rarity,
        string mathType, bool hovered = false)
    {
        Color rarityColor = UiTheme.RarityColors.TryGetValue(rarity, out var color) ? color : UiTheme.Border;
        int cornerRadius = UiTheme.CardCornerRadius(rect.Width);
        var shadow = new Rectangle(rect.X + Math.Max(2, rect.Width / 12), rect.Y + Math.Max(2, rect.Width / 12), rect.Width, rect.Height);
        Primitives2D.FillRoundedRect(spriteBatch, shadow, UiTheme.Shadow, cornerRadius);
        Color fill = hovered ? UiTheme.Lighten(rarityColor, 24) : rarityColor;
        Primitives2D.FillRoundedRect(spriteBatch, rect, fill, cornerRadius);
        Primitives2D.RoundedRectOutline(spriteBatch, rect, UiTheme.Ink, Math.Max(2, rect.Width / 14), cornerRadius);
        var inner = rect;
        inner.Inflate((int)(-rect.Width * .18f), (int)(-rect.Height * .22f));
        DrawStatSymbol(spriteBatch, statName, inner, UiTheme.Ink);

        var markCenter = new Vector2(rect.Right - rect.Width * .18f, rect.Y + rect.Height * .17f);
        float mark = rect.Width * .10f;
        int stroke = Math.Max(2, rect.Width / 14);
        if (mathType == "multiplicative")
        {
            DrawLine(spriteBatch, UiTheme.Text, markCenter - new Vector2(mark, mark), markCenter + new Vector2(mark, mark), stroke);
            DrawLine(spriteBatch, UiTheme.Text, markCenter + new Vector2(mark, -mark), markCenter + new Vector2(-mark, mark), stroke);
        }
        else
        {
            DrawLine(spriteBatch, UiTheme.Text, markCenter - new Vector2(mark, 0), markCenter + new Vector2(mark, 0), stroke);
            DrawLine(spriteBatch, UiTheme.Text, markCenter - new Vector2(0, mark), markCenter + new Vector2(0, mark), stroke);
        }
        return rect;
    }
}
